### PR TITLE
Replace _dynamo.config with an object instead of module

### PR DESCRIPTION
**Pull request by qihqi**
Summary:
    Replace _dynamo.config with an object instead of module

    Current usage patterns of setting and reading fields on config will work
    unchanged.

    Only changes needed going forward:
    1. import torch._dynamo.config will not work. However, just doing
       import torch._dynamo is sufficient to access dynamo config
       as torch._dynamo.config.

    2. Files inside of _dynamo folder need to access config via
       from torch._dynamo.config_util import config instead of
       from torch._dynamo import config. Because _dynamo/__init__.py
       imports some of the files so it would be circular import.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Fixes #ISSUE_NUMBER

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Comment by yanbing-j** [comment]
Hi @qihqi , I suppose the `third_party/ideep` change in this PR is not intentional. Could you please help double check and remove the change? Thank you!

**Comment by qihqi** [comment]
> 

Yes. Removed.

File: torch/_dynamo/_config.py
Comment (williamwen42) [inline torch/_dynamo/_config.py:1]: Is this module still needed?

**Review by williamwen42 — COMMENTED** [review]
Will you also do the same for inductor config?

File: torch/_dynamo/_config.py
Comment (qihqi) [reply torch/_dynamo/_config.py:1]: no. removed.

Comment (qihqi) [reply torch/_dynamo/_config.py:1]: Probly i'll do the same for inductor config; but not the same PR

**Review by williamwen42 — DISMISSED** [review]

**Comment by voznesenskym** [comment]
Why did we make this change?

**Comment by jansel** [comment]
This will break a lot of user code (we have been telling people about dynamo.config), with seemingly little benefit.

**Review by jansel — CHANGES_REQUESTED** [review]
This will break lot of user code with little benefit.  If there is a good reason for this change it should be done in a non-BC breaking way.

**Review by jansel — CHANGES_REQUESTED** [review]

File: test/inductor/test_config.py [hunk 1]
```diff
@@ -28,8 +28,6 @@ def test_set(self):
         config.max_fusion_size = 13337
         self.assertEqual(config.max_fusion_size, 13337)
         self.assertEqual(config.to_dict()["max_fusion_size"], 13337)
-        config.to_dict()["max_fusion_size"] = 32
```
[the remark this answers is not in the document]
Comment (qihqi) [reply test/inductor/test_config.py:31]: Which places? I can fix up instead?

The thing is, with a dataclass the data is not backend by a flatten'd dict

[the remark this answers is not in the document]
Comment (qihqi) [reply test/inductor/test_config.py:31]: Grepping didnt yield any related things:
(torch39) han@helene ~/code/pytorch $ grep 'to_dict()' torch -r
torch/_dynamo/config_utils.py:        self.to_dict().update(pickle.loads(data))
torch/_inductor/__init__.py:    current_config: Dict[str, Any] = config.to_dict()  # type: ignore[attr-defined]
torch/__init__.py:        current_config: Dict[str, Any] = config.to_dict()  # type: ignore[attr-defined]
(plus things under quantization/ which is probly a different config).

File: torch/_dynamo/__init__.py [hunk 2]
```diff
@@ -1,3 +1,17 @@
...
+
+# Class type of config
+# torch._dynamo.config is really an object of this type instead of a module.
+from . import config as _config_module
+
+DynamoConfig = torch._config_utils.make_config_dataclass("DynamoConfig", _config_module)
+DynamoConfig.__doc__ = "Dataclass that holds configs related to dynamo."
+config = DynamoConfig()
+config.__name__ = f"{__name__}.config"
+sys.modules[f"{__name__}.config"] = config
```
[the remark this answers is not in the document]
Comment (qihqi) [reply torch/_dynamo/__init__.py]: sure. done

File: test/inductor/test_config.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jansel) [reply test/inductor/test_config.py:31]: Not all the users are going to be in the pytorch codebase.  One of the side effects of being an open source project.  This seems like something that is BC breaking for no reason.  And will also result in really nasty bugs for anyone relying on this since the deleted test will run fine without error, but silently do nothing.

[the remark this answers is not in the document]
Comment (qihqi) [reply test/inductor/test_config.py]: Sure. done.

**Comment by qihqi** [comment]
Hi @jansel any more comments? Looks like the CIs is also looking good. Thanks!

File: torch/_dynamo/allowed_functions.py [hunk 3]
```diff
@@ -164,7 +164,7 @@ def _is_allowed_module_prefix(obj):
         if module is None:
             return False
 
-        mod_name = module.__name__
+        mod_name = getattr(module, "__name__", "")
```
Comment (jansel) [inline torch/_dynamo/allowed_functions.py:167]: Why is this change needed?

File: torch/_dynamo/__init__.py [hunk 4]
```diff
@@ -1,3 +1,14 @@
+import sys
+
+import torch._config_utils
```
Comment (jansel) [inline torch/_dynamo/__init__.py:3]: Unused?

File: torch/_dynamo/__init__.py [hunk 5]
```diff
@@ -1,3 +1,14 @@
+import sys
```
Comment (jansel) [inline torch/_dynamo/__init__.py:1]: Unused?

File: torch/_dynamo/config.py [hunk 6]
```diff
@@ -14,30 +15,30 @@
...
 
 # whether or not to specialize on int inputs.  This only has an effect with
 # dynamic_shapes; when dynamic_shapes is False, we ALWAYS specialize on int
 # inputs
-specialize_int = False
+specialize_int: bool = False
 
 # Assume these functions return constants
-constant_functions = {
+constant_functions: Dict[ModuleType, bool] = {
```
Comment (jansel) [inline torch/_dynamo/config.py:41]: This type is wrong, these are functions not modules

File: torch/_dynamo/config.py [hunk 7]
```diff
@@ -176,63 +179,63 @@
...
+error_on_recompile: bool = False
 
 # reports why guards fail. Useful to identify the guards failing frequently and
 # causing recompilations.
-report_guard_failures = os.environ.get("TORCHDYNAMO_REPORT_GUARD_FAILURES") == "1"
+report_guard_failures: bool = os.environ.get("TORCHDYNAMO_REPORT_GUARD_FAILURES") == "1"
 
 # root folder of the project
-base_dir = dirname(dirname(dirname(abspath(__file__))))
+base_dir: bool = dirname(dirname(dirname(abspath(__file__))))
```
Comment (jansel) [inline torch/_dynamo/config.py:235]: Type is wrong here too.

File: torch/_dynamo/config.py [hunk 8]
```diff
@@ -133,25 +135,26 @@
...
     "torch._decomp",
 }
 
+
 # Debug Flag to try minifier at different stages. Possible values are {None, "aot", "dynamo"}
 # None - Minifier is switched off
 # dynamo - Runs minifier on the TorchDynamo produced graphs, if compilation fails
 # aot - Runs minifier on the Aot Autograd produced graphs, if compilation fails
-repro_after = os.environ.get("TORCHDYNAMO_REPRO_AFTER", None)
+repro_after: bool = os.environ.get("TORCHDYNAMO_REPRO_AFTER", None)
```
Comment (jansel) [inline torch/_dynamo/config.py:151]: Type is wrong

File: torch/_functorch/__init__.py [hunk 9]
```diff
@@ -3,3 +3,9 @@
 #
 # This source code is licensed under the BSD-style license found in the
 # LICENSE file in the root directory of this source tree.
+import sys
+from .._config_utils import make_config_dataclass
+from . import config
+
+FunctorchConfig = type(config)
```
Comment (jansel) [inline torch/_functorch/__init__.py:10]: Should this be FuncTorchConfig?

File: torch/_inductor/__init__.py [hunk 10]
```diff
@@ -1,8 +1,14 @@
+import sys
```
Comment (jansel) [inline torch/_inductor/__init__.py:1]: Unused?

File: torch/_inductor/__init__.py [hunk 11]
```diff
@@ -1,8 +1,14 @@
...
 from typing import Any, Dict, List, Optional
 
 import torch.fx
+from .._config_utils import make_config_dataclass
```
Comment (jansel) [inline torch/_inductor/__init__.py:5]: Unused?

File: torch/_inductor/config.py [hunk 12]
```diff
@@ -1,129 +1,137 @@
...
+realize_reads_threshold: int = 4
+realize_bytes_threshold: int = 2000
 
 # Threshold to prevent excessive accumulation of ops in one buffer during lowering
-realize_acc_reads_threshold = 8
+realize_acc_reads_threshold: int = 8
 
 # fallback to eager for random/dropout, this is slow but useful for debugging
-fallback_random = False
+fallback_random: int = False
```
Comment (jansel) [inline torch/_inductor/config.py:91]: Wrong type

File: torch/_inductor/config.py [hunk 13]
```diff
@@ -1,129 +1,137 @@
...
 
 # automatically create fallbacks when encountering an unhandled op
-implicit_fallbacks = True
+implicit_fallbacks: int = True
```
Comment (jansel) [inline torch/_inductor/config.py:94]: Wrong type

File: torch/_inductor/config.py [hunk 14]
```diff
@@ -1,129 +1,137 @@
...
 
 # fuse even in cases without common reads
-aggressive_fusion = False
+aggressive_fusion: int = False
```
Comment (jansel) [inline torch/_inductor/config.py:97]: Wrong type

File: torch/_inductor/config.py [hunk 15]
```diff
@@ -1,129 +1,137 @@
...
+benchmark_kernel: bool = os.environ.get("TORCHINDUCTOR_BENCHMARK_KERNEL", "0") == "1"
 
 
 def is_fbcode():
     return not hasattr(torch.version, "git_version")
 
 
 # warnings intended for PyTorch developers, disable for point releases
-is_nightly_or_source = "dev" in torch.__version__ or "git" in torch.__version__
+is_nightly_or_source: str = "dev" in torch.__version__ or "git" in torch.__version__
```
Comment (jansel) [inline torch/_inductor/config.py:126]: Wrong type

File: torch/_inductor/config.py [hunk 16]
```diff
@@ -141,64 +149,67 @@ def decide_compile_threads():
         )
 
 
-compile_threads = decide_compile_threads()
+def is_fbcode():
```
Comment (jansel) [inline torch/_inductor/config.py:152]: There is also a version of this in ConfigMixin.  Which one gets called?  Could lead to confusing bugs if one edits the wrong one.

File: torch/_inductor/config.py [hunk 17]
```diff
@@ -208,150 +219,153 @@ class cpp:
...
     # Upload the .tar.gz file
     # Needs to be overriden based on specific environment needs
-    upload_tar = None
-
+    upload_tar: Any = dataclasses.field(default=None, metadata={"skip_pickle": True})
 
-_save_config_ignore = {
-    # workaround: "Can't pickle <function ...>"
-    "trace.upload_tar",
-}
```
Comment (jansel) [inline torch/_inductor/config.py:351]: Won't we still need this workaround?  I believe this was an fbcode issue.

**Review by jansel — CHANGES_REQUESTED** [review]
Please double check types.  I found a lot of incorrect ones.

Please add tests, with added complexity I am worried some of the function (especially around nested configs) won't work right and arenn't being tested.

**Review by qihqi — COMMENTED** [review]
Double checked

**Review by qihqi — COMMENTED** [review]
Double checked

File: torch/_inductor/config.py (hunk 17 above)
Comment (qihqi) [reply torch/_inductor/config.py:357]: pickle will try to pickle the `trace` object, which will then call `__getstate__` on it, `__getstate__` defined by ConfigMixin will ignore fields marked with skip_pickle.

I added a test to make sure.

**Review by qihqi — COMMENTED** [review]
Added unit test and docstring for ConfigMixin

File: torch/_dynamo/config.py [hunk 18]
```diff
@@ -16,30 +18,30 @@
...
 # inputs
-specialize_int = False
+specialize_int: bool = False
 
 # Assume these functions return constants
-constant_functions = {
+constant_functions: Dict[Callable, bool] = {
     torch.jit.is_scripting: False,
     torch.jit.is_tracing: False,
     torch._C._get_tracing_state: None,
```
Comment (williamwen42) [inline torch/_dynamo/config.py:47]: Type mismatch between `None` and `bool`

File: torch/_inductor/config.py [hunk 19]
```diff
@@ -211,153 +214,157 @@ class cpp:
...
+    inject_relu_bug_TESTING_ONLY: Optional[str] = None
+    inject_log1p_bug_TESTING_ONLY: Optional[str] = None
 
     # If None, autodetect whether or not AVX512/AVX2 can be used.  Otherwise,
     # force usage as specified, without testing.
-    vec_isa_ok = None
+    vec_isa_ok: Optional[str] = None
 
     # similar to config.triton.descriptive_names
     descriptive_names = "original_aten"
```
Comment (williamwen42) [inline torch/_inductor/config.py:233]: Missing annotation?

**Review by williamwen42 — COMMENTED** [review]
LGTM other than a few nits, deferring final approval to @jansel.

File: torch/_dynamo/config.py (hunk 18 above)
Comment (qihqi) [reply torch/_dynamo/config.py:49]: done.

**Review by jansel — CHANGES_REQUESTED** [review]
Please remove the changes to `third_party/kineto`

**Comment by qihqi** [comment]
> Please remove the changes to `third_party/kineto`

Done.

**Comment by jansel** [comment]
Tests seem broken.

**Review by jansel — APPROVED** [review]

**Comment by qihqi** [comment]
Thanks Jason!

Hi @williamwen42 can you delink the Meta diff with this PR so the meta internal changes check can pass? 
I'll figure out the EasyCLA thingy meanwhile. Thanks!

**Comment by williamwen42** [comment]
Unlinked the PR, but the meta internal check is still failing. Looped in with devinfra and was advised that this behavior is strange and that we can just merge -f once the CLA check passes.

**Comment by DanilBaibak** [comment]
Hi @qihqi! There is a tiny bug in the changes. You [removed](https://github.com/pytorch/pytorch/pull/96455/files#diff-5f3e7e4866f93af05691e2382d2258cad959091492f900583a49e55266f5e897L5) the `tempfile` import, but it [still used in the code](https://github.com/pytorch/pytorch/pull/96455/files#diff-5f3e7e4866f93af05691e2382d2258cad959091492f900583a49e55266f5e897L249).

**Comment by qihqi** [comment]
> Hi @qihqi! There is a tiny bug in the changes. You [removed](https://github.com/pytorch/pytorch/pull/96455/files#diff-5f3e7e4866f93af05691e2382d2258cad959091492f900583a49e55266f5e897L5) the `tempfile` import, but it [still used in the code](https://github.com/pytorch/pytorch/pull/96455/files#diff-5f3e7e4866f93af05691e2382d2258cad959091492f900583a49e55266f5e897L249).

good catch! sorry about that
